test/test_rrules.jl

```diff
@@ -0,0 +1,111 @@
+w = GenSimSourceMulti(xsrc_index, zsrc_index, nsrc, model.n);
+
+sinput = zip(["Point", "Extended"], [Ps, Pw], (q, w))
+#####################################################################################
+ftol = sqrt(eps(1f0))
+
+@testset "AD correctness check return_array=$(ra)" for ra in [true, false]
+    opt = Options(return_array=ra, sum_padding=true, f0=f0)
+    A_inv = judiModeling(model; options=opt)
+    A_inv0 = judiModeling(model0; options=opt)
+    @testset "AD correctness check source type: $(stype)" for (stype, Pq, q) in sinput
+        @timeit TIMEROUTPUT "$(stype) source AD, array=$(ra)" begin
+            # Linear operators
+            q0 = perturb(q)
+            # Operators
+            F = Pr*A_inv*adjoint(Pq)
+            F0 = Pr*A_inv0*adjoint(Pq)
+
+            d_obs = F(m, q)
+            # PDE accept model as input but AD expect the actual model param (model.m)
+            d_obs2 = F(model, q)
+            @test d_obs ≈ d_obs2 rtol=ftol
+
+            J = judiJacobian(F0, q0)
+            gradient_m = adjoint(J)*(F(m0, q0) - d_obs)
+            gradient_m2 = adjoint(J)*(F(model0.m, q0) - d_obs)
+            @test gradient_m ≈ gradient_m2 rtol=ftol
+
+            # Gradient with m array
+            gs_inv = gradient(x -> misfit_objective(d_obs, q0, x, F), m0)
+            # Gradient with m PhysicalParameter
+            gs_inv2 = gradient(x -> misfit_objective(d_obs, q0, x, F), model0.m)
+            @test gs_inv[1][:] ≈ gs_inv2[1][:] rtol=ftol
+
+            g1 = vec(gradient_m)
+            g2 = vec(gs_inv[1])
+
+            @test isapprox(norm(g1 - g2) / norm(g1 + g2), 0f0; atol=ftol)
+            @test isapprox(dot(g1, g2)/norm(g1)^2,1f0;rtol=ftol)
+            @test isapprox(dot(g1, g2)/norm(g2)^2,1f0;rtol=ftol)
+        end
+    end
+end
+
+
+@testset "AD Gradient test return_array=$(ra)" for ra in [false, true]
+    opt = Options(return_array=ra, sum_padding=true, f0=f0, dt_comp=dt)
+    F = judiModeling(model; options=opt)
+    ginput = zip(["Point", "Extended"], [Pr*F*Ps', Pr*F*Pw'], (q, w))
+    @testset "Gradient test: $(stype) source" for (stype, F, q) in ginput
+        @timeit TIMEROUTPUT "$(stype) source gradient, array=$(ra)" begin
+            # Initialize source for source perturbation
+            q0 = perturb(q)
+            # Data and source perturbation
+            d, dq = F*q, q-q0
+
+            #####################################################################################
+            f0, gq, gm = loss(d, q0, m0, F)
+            # Gradient test for extended modeling: source
+            print("\nGradient test source $(stype) source, array=$(ra)\n")
+            grad_test(x-> misfit_objective(d, x, m0, F), q0, dq, gq)
+  
+            # Gradient test for extended modeling: model
+            print("\nGradient test model $(stype) source, array=$(ra)\n")
+            grad_test(x-> misfit_objective(d, q0, x, F), m0, dm, gm)
```

Comment:
If you put this test in test_gradient.jl then you save a lot of PDE solves because you don't need to run all iterations in gradient test. But this looks clean. Up to you.